src/Analysis/Engine/Impl/Infrastructure/TextBuilder.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.PythonTools.Analysis.Infrastructure {
+    internal class TextBuilder {
+        private readonly List<string> _segments = new List<string>();
+
+        public override string ToString() {
+            IEnumerable<string> seq = _segments;
+            if (IsLastWhitespace()) {
+                // Don't mutate when producing the string.
+                seq = _segments.Take(_segments.Count - 1);
+            }
+            return string.Concat(seq);
+        }
+
+        public void Append(string s) => _segments.Add(s);
+
+        public void Append(object o) => _segments.Add(o.ToString());
```

Comment:
I think string is kind of object, so string-specific one may not be necessary

Reply:
I removed the string specialization.